{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b9e2b39d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pyspark in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from -r requirements.txt (line 1)) (3.5.0)\n",
      "Requirement already satisfied: pandas in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from -r requirements.txt (line 2)) (2.2.0)\n",
      "Requirement already satisfied: numpy in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from -r requirements.txt (line 3)) (1.26.4)\n",
      "Requirement already satisfied: polars in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from -r requirements.txt (line 5)) (0.20.7)\n",
      "Requirement already satisfied: openpyxl in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from -r requirements.txt (line 6)) (3.1.2)\n",
      "Requirement already satisfied: tqdm in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from -r requirements.txt (line 7)) (4.66.1)\n",
      "Requirement already satisfied: sqlalchemy in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from -r requirements.txt (line 8)) (2.0.25)\n",
      "Requirement already satisfied: python-dotenv in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from -r requirements.txt (line 9)) (1.0.1)\n",
      "Requirement already satisfied: psycopg2 in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from -r requirements.txt (line 10)) (2.9.9)\n",
      "Requirement already satisfied: py4j==0.10.9.7 in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from pyspark->-r requirements.txt (line 1)) (0.10.9.7)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from pandas->-r requirements.txt (line 2)) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from pandas->-r requirements.txt (line 2)) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from pandas->-r requirements.txt (line 2)) (2023.4)\n",
      "Requirement already satisfied: et-xmlfile in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from openpyxl->-r requirements.txt (line 6)) (1.1.0)\n",
      "Requirement already satisfied: typing-extensions>=4.6.0 in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from sqlalchemy->-r requirements.txt (line 8)) (4.9.0)\n",
      "Requirement already satisfied: greenlet!=0.4.17 in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from sqlalchemy->-r requirements.txt (line 8)) (3.0.3)\n",
      "Requirement already satisfied: six>=1.5 in /home/jaikar/genepower/.venv/lib/python3.12/site-packages (from python-dateutil>=2.8.2->pandas->-r requirements.txt (line 2)) (1.16.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install -r requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "17acc02f-2ea8-4375-a411-79fd5895a851",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "24/02/15 11:11:34 WARN Utils: Your hostname, user1-PowerEdge-T40 resolves to a loopback address: 127.0.1.1; using 10.4.16.223 instead (on interface eno1)\n",
      "24/02/15 11:11:34 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address\n",
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "24/02/15 11:11:34 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n",
      "24/02/15 11:11:35 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.\n"
     ]
    }
   ],
   "source": [
    "from pyspark import SparkContext\n",
    "from pyspark.sql import SparkSession\n",
    "import pandas as pd\n",
    "spark = SparkSession.builder \\\n",
    "    .appName('Read CSV File into DataFrame') \\\n",
    "    .config('spark.executor.memory', '20g') \\\n",
    "    .config('spark.driver.memory', '32g') \\\n",
    "    .getOrCreate()\n",
    "# spark = SparkSession.builder.appName('Read CSV File into DataFrame').getOrCreate()\n",
    "from pyspark.sql.functions import col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9f19d761-a5fc-459a-aab4-13db949ac265",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded required Libraries\n"
     ]
    }
   ],
   "source": [
    "##################################### IMPORTING THE REQUIRED LIBRARIES ##########################################################\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "import re\n",
    "import os\n",
    "pd.set_option('display.max_columns',None)\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "print(\"Loaded required Libraries\")\n",
    "\n",
    "######################################### GETTING THE COVERED POSITIONS FROM THE MOTHER VCF FILE FROM THE BED COORDINATES ####\n",
    "\n",
    "def read_bed_file(bed_file):\n",
    "    bed_positions = set()\n",
    "    with open(bed_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('#'):  # Skip header lines if present\n",
    "                continue\n",
    "            fields = line.strip().split('\\t')\n",
    "            if len(fields) >= 3:\n",
    "                chrom = fields[0]\n",
    "                try:\n",
    "                    start = int(fields[1])\n",
    "                    end = int(fields[2])\n",
    "                except ValueError:\n",
    "                    continue  # Skip this line if start or end position is not an integer\n",
    "                for pos in range(start, end + 1):\n",
    "                    bed_positions.add((chrom, pos))\n",
    "    return bed_positions\n",
    "\n",
    "def normalize_chrom_name(chrom):\n",
    "    return chrom.split('_')[0]\n",
    "\n",
    "def filter_vcf_file(vcf_file, bed_positions):\n",
    "    filtered_vcf_records = []\n",
    "    with open(vcf_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('#'):  # Preserve header lines in the output\n",
    "                filtered_vcf_records.append(line)\n",
    "                continue\n",
    "            fields = line.strip().split('\\t')\n",
    "            if len(fields) >= 2:\n",
    "                raw_chrom = fields[0]\n",
    "                chrom = normalize_chrom_name(raw_chrom)\n",
    "                try:\n",
    "                    pos = int(fields[1])\n",
    "                except ValueError:\n",
    "                    continue  # Skip this line if 'POS' is not an integer\n",
    "                if (chrom, pos) in bed_positions:\n",
    "                    filtered_vcf_records.append(line)\n",
    "    return filtered_vcf_records\n",
    "\n",
    "def write_filtered_vcf(filtered_vcf_records, output_file):\n",
    "    with open(output_file, 'w') as f:\n",
    "        for record in filtered_vcf_records:\n",
    "            f.write(record)\n",
    "bed_positions = read_bed_file(r'Covered_regions.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "65e104e0-f707-479b-85ab-a45ae7f4d626",
   "metadata": {},
   "outputs": [],
   "source": [
    "    ########################################### IMPORTING THE VCF DATA AND EXPANDING THE DEPTH COLUMNS ##########################\n",
    "def process(filepath,output_dir):\n",
    "        filename= os.path.basename(filepath)\n",
    "        filename,_ = os.path.splitext(filename)\n",
    "        filtered_vcf = filter_vcf_file(filepath, bed_positions)\n",
    "        write_filtered_vcf(filtered_vcf, filepath)\n",
    "        vcf = pd.read_csv(filepath, comment= '#', sep = '\\t', header=None, low_memory=False)\n",
    "        vcf.columns = ['CHROM', 'POS', 'rsID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'SAMPLE']\n",
    "\n",
    "        sample_cols = vcf['SAMPLE'].str.split(':', expand=True)\n",
    "        sample_cols.columns = ['GT', 'GQ', 'SDP', 'DP', 'RD', 'AD', 'FREQ', 'PVAL', 'RBQ', 'ABQ', 'RDF', 'RDR', 'ADF', 'ADR']\n",
    "\n",
    "        # Assign the values to the newly created columns\n",
    "        vcf = pd.concat([vcf, sample_cols], axis=1)\n",
    "        vcf = vcf[['CHROM', 'POS', 'rsID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'GT', 'GQ', 'SDP', 'RBQ','ABQ','DP', 'RD', 'AD', 'FREQ', 'PVAL','RDF', 'RDR', 'ADF', 'ADR']]\n",
    "\n",
    "        print('Loading the Data completed and Depth columns splitted')\n",
    "\n",
    "        ######################################### EXTRACTING THE ZYGOSITY FROM THE INFO COLUMN OF THE EACH VARIANT ##############\n",
    "\n",
    "        vcf['HET'] = vcf['INFO'].str.extract(r'HET=(\\d)')\n",
    "        vcf['HOM'] = vcf['INFO'].str.extract(r'HOM=(\\d)')\n",
    "\n",
    "        # Create a new column 'Zygosity' based on conditions\n",
    "        vcf['Zygosity'] = ''\n",
    "\n",
    "        vcf.loc[vcf['HOM'] == '1', 'Zygosity'] = 'Homozygous'\n",
    "        vcf.loc[vcf['HET'] == '1', 'Zygosity'] = 'Heterozygous'\n",
    "        vcf['GT'] = vcf['GT'].astype(str)\n",
    "\n",
    "        print('Zygosity extraction completed')\n",
    "\n",
    "        ######################################## EXTRACTING THE GENEINFO FROM THE INFO COLUMN ####################################\n",
    "\n",
    "        vcf[\"Gene_Name\"] = vcf[\"INFO\"].str.extract('GENEINFO=(?P<GENEINFO>.+?);')\n",
    "        vcf['Gene Name'] = vcf['Gene_Name'].apply(lambda x: ','.join([segment.split(':')[0] for segment in x.split('|')]) if pd.notnull(x) else '')\n",
    "\n",
    "        print('Gene extraction completed')\n",
    "\n",
    "        ####################################### SPLITTING AND EXPLODING THE CSQ COLUMN FOR THE REQUIRED PARAMETERS ##############\n",
    "\n",
    "        vcf['CSQ'] = vcf['INFO'].str.extract(r'CSQ=(.*)')\n",
    "        vcf['csq'] = vcf['CSQ'].str.split(',')\n",
    "        vcf = vcf.explode('csq')\n",
    "\n",
    "        print('CSQ splitting completed')\n",
    "\n",
    "        ###################################### EXTRACTION OF THE REQUIRED KEY-VALUE PAIRS FROM THE CSQ ##########################\n",
    "\n",
    "        vcf['ClinVar_CLNDN'] = vcf['csq'].str.split('|').str[82]\n",
    "        vcf['CLIN_SIG'] = vcf['csq'].str.split('|').str[70]\n",
    "        vcf['ClinVar_CLNREVSTAT'] = vcf['csq'].str.split('|').str[81]\n",
    "        vcf['ClinVar'] = vcf['csq'].str.split('|').str[79]\n",
    "        vcf['HGVSC'] = vcf['csq'].str.split('|').str[10]\n",
    "        vcf['HGVSP'] = vcf['csq'].str.split('|').str[11]\n",
    "        vcf['PolyPhen'] = vcf['csq'].str.split('|').str[38]\n",
    "        vcf['BIOTYPE'] = vcf['csq'].str.split('|').str[7]\n",
    "        vcf['EXON'] = vcf['csq'].str.split('|').str[8]\n",
    "        vcf['INTRON'] = vcf['csq'].str.split('|').str[9]\n",
    "        vcf['Protein_position'] = vcf['csq'].str.split('|').str[14]\n",
    "        vcf['Amino_acids'] = vcf['csq'].str.split('|').str[15]\n",
    "        vcf['Codons'] = vcf['csq'].str.split('|').str[16]\n",
    "        vcf['STRAND'] = vcf['csq'].str.split('|').str[19]\n",
    "        vcf['PUBMED'] = vcf['csq'].str.split('|').str[73]\n",
    "        vcf['Consequence'] = vcf['csq'].str.split('|').str[1]\n",
    "        vcf['IMPACT'] = vcf['csq'].str.split('|').str[2]\n",
    "        vcf['SIFT'] = vcf['csq'].str.split('|').str[37]\n",
    "        vcf['AF'] = vcf['csq'].str.split('|').str[42]\n",
    "        vcf['AFR_AF'] = vcf['csq'].str.split('|').str[43]\n",
    "        vcf['AMR_AF'] = vcf['csq'].str.split('|').str[44]\n",
    "        vcf['EAS_AF'] = vcf['csq'].str.split('|').str[45]\n",
    "        vcf['EUR_AF'] = vcf['csq'].str.split('|').str[46]\n",
    "        vcf['SAS_AF'] = vcf['csq'].str.split('|').str[47]\n",
    "        vcf['gnomADe_AF'] = vcf['csq'].str.split('|').str[48]\n",
    "        vcf['gnomADe_AFR_AF'] = vcf['csq'].str.split('|').str[49]\n",
    "        vcf['gnomADe_AMR_AF'] = vcf['csq'].str.split('|').str[50]\n",
    "        vcf['gnomADe_ASJ_AF'] = vcf['csq'].str.split('|').str[51]\n",
    "        vcf['gnomADe_EAS_AF'] = vcf['csq'].str.split('|').str[52]\n",
    "        vcf['gnomADe_FIN_AF'] = vcf['csq'].str.split('|').str[53]\n",
    "        vcf['gnomADe_NFE_AF'] = vcf['csq'].str.split('|').str[54]\n",
    "        vcf['gnomADe_OTH_AF'] = vcf['csq'].str.split('|').str[55]\n",
    "        vcf['gnomADe_SAS_AF'] = vcf['csq'].str.split('|').str[56]\n",
    "        vcf['gnomADg_AF'] = vcf['csq'].str.split('|').str[57]\n",
    "        vcf['gnomADg_AFR_AF'] = vcf['csq'].str.split('|').str[58]\n",
    "        vcf['gnomADg_AMI_AF'] = vcf['csq'].str.split('|').str[59]\n",
    "        vcf['gnomADg_AMR_AF'] = vcf['csq'].str.split('|').str[60]\n",
    "        vcf['gnomADg_ASJ_AF'] = vcf['csq'].str.split('|').str[61]\n",
    "        vcf['gnomADg_EAS_AF'] = vcf['csq'].str.split('|').str[62]\n",
    "        vcf['gnomADg_FIN_AF'] = vcf['csq'].str.split('|').str[63]\n",
    "        vcf['gnomADg_MID_AF'] = vcf['csq'].str.split('|').str[64]\n",
    "        vcf['gnomADg_NFE_AF'] = vcf['csq'].str.split('|').str[65]\n",
    "        vcf['gnomADg_OTH_AF'] = vcf['csq'].str.split('|').str[66]\n",
    "        vcf['gnomADg_SAS_AF'] = vcf['csq'].str.split('|').str[67]\n",
    "        vcf['MAX_AF'] = vcf['csq'].str.split('|').str[68]\n",
    "        vcf['MAX_AF_POPS'] = vcf['csq'].str.split('|').str[69]\n",
    "\n",
    "        print('Required columns extraction completed')\n",
    "        ############################################### Protein Position and Amino Acid Change ##################################\n",
    "        vcf['Protein Position and Amino Acid'] = vcf['Amino_acids'].str[0] + vcf['Protein_position'] + np.where(vcf['Amino_acids'].str[-1] == vcf['Amino_acids'].str[0], '', vcf['Amino_acids'].str[-1])\n",
    "\n",
    "        ############################################### HGVSc AND HGVSp TRANSCRIPTS EXTRACTION ###################################\n",
    "\n",
    "        vcf[['HGVSc', 'HGVSc (Transcript)']] = vcf['HGVSC'].str.split(':' ,expand=True)\n",
    "        vcf[['HGVSp', 'HGVSp (Transcript)']] = vcf['HGVSP'].str.split(':', expand=True)\n",
    "        vcf_final = vcf.copy()\n",
    "\n",
    "        print('Protein_HGVSc_HGVSp_extraction completed')\n",
    "\n",
    "        ############################################### REMOVING THE [\"not_specified\", \"not_provided\"] FROM THE COLUMNS ##########\n",
    "\n",
    "        # Define the terms to remove\n",
    "        remove_terms = set([\"not_specified\", \"not_provided\"])\n",
    "\n",
    "        # Apply the filtering operation to 'Column1' only\n",
    "        vcf_final['ClinVar_CLNDN'] = vcf_final['ClinVar_CLNDN'].apply(lambda row: \"&\".join(\n",
    "            [term for term in row.split(\"&\") if term not in remove_terms]\n",
    "            ) if isinstance(row, str) and not all(term in remove_terms for term in row.split(\"&\")) else row)\n",
    "\n",
    "\n",
    "        vcf_final['CLIN_SIG'] = vcf_final['CLIN_SIG'].apply(lambda row: \"&\".join(\n",
    "            [term for term in row.split(\"&\") if term not in remove_terms]\n",
    "            ) if isinstance(row, str) and not all(term in remove_terms for term in row.split(\"&\")) else row)\n",
    "\n",
    "\n",
    "        vcf_final['ClinVar_CLNREVSTAT'] = vcf_final['ClinVar_CLNREVSTAT'].apply(lambda row: \"&\".join(\n",
    "            [term for term in row.split(\"&\") if term not in remove_terms]\n",
    "            ) if isinstance(row, str) and not all(term in remove_terms for term in row.split(\"&\")) else row)\n",
    "\n",
    "        print('\"not_specified\", \"not_provided\" completed')\n",
    "\n",
    "        # Assuming you already have a DataFrame named vcf_final\n",
    "        replace_dict = {'&': ',', '_': ' '}\n",
    "\n",
    "        # Set the chunk size based on your available memory\n",
    "        chunk_size = 1000\n",
    "\n",
    "        # Get the number of chunks\n",
    "        num_chunks = int(np.ceil(len(vcf_final) / chunk_size))\n",
    "\n",
    "        # Create an empty list to store the processed chunks\n",
    "        processed_chunks = []\n",
    "\n",
    "        # Iterate over the DataFrame in chunks\n",
    "        for chunk_number in range(num_chunks):\n",
    "            start_index = chunk_number * chunk_size\n",
    "            end_index = min((chunk_number + 1) * chunk_size, len(vcf_final))\n",
    "            \n",
    "            # Get the current chunk\n",
    "            current_chunk = vcf_final.iloc[start_index:end_index]\n",
    "            \n",
    "            # Replace characters in the current chunk\n",
    "            current_chunk = current_chunk.astype(str).replace(replace_dict, regex=True)\n",
    "            \n",
    "            # Perform additional processing on the chunk if needed\n",
    "            \n",
    "            # Append the processed chunk to the list\n",
    "            processed_chunks.append(current_chunk)\n",
    "\n",
    "        # Concatenate all processed chunks into the final DataFrame\n",
    "        vcf_final = pd.concat(processed_chunks, ignore_index=True)\n",
    "\n",
    "        vcf_final['consequence'] = vcf_final['Consequence'].str.split(',').str[0]\n",
    "\n",
    "        df_1 = pd.read_excel(r'consequence.xlsx')\n",
    "\n",
    "        merged_1 = pd.merge(vcf_final, df_1, on='consequence', how='left', sort=False)\n",
    "\n",
    "        df_2 = pd.read_excel(r'IMPACT.xlsx')\n",
    "\n",
    "        merged_2 = pd.merge(merged_1, df_2, on = 'IMPACT', how='left', sort=False)\n",
    "\n",
    "        print('Scores added')\n",
    "\n",
    "        ############################################# CONDITION GENES MAPPING TO THE MAIN VCF ######################################\n",
    "\n",
    "        df_gene = pd.read_excel(r'Conditions_final_genes.xlsx')\n",
    "\n",
    "        merged_2['Gene Match'] = 'No'\n",
    "        merged_2['Matched_Gene'] = ''\n",
    "            \n",
    "        # Iterate through each gene in vcf['Gene']\n",
    "        for index, genes in merged_2['Gene Name'].items():\n",
    "            if isinstance(genes, str):\n",
    "                gene_list = genes.split(',')\n",
    "                for gene in gene_list:\n",
    "                    if gene in df_gene['Gene Name'].values:\n",
    "                        merged_2.at[index, 'Gene Match'] = 'Yes'\n",
    "                        merged_2.at[index, 'Matched_Gene'] = gene\n",
    "                        break\n",
    "            \n",
    "        df_gene = df_gene.rename({'Gene Name':'Matched_Gene'}, axis=1)\n",
    "\n",
    "        #merged_2 = merged_2.drop(columns=['Gene Match'], axis=1)\n",
    "\n",
    "        merged_2 = pd.merge(merged_2, df_gene, on= 'Matched_Gene', how = 'left', sort = False)\n",
    "        merged_2['Condition'] = merged_2['Condition'].fillna('No')\n",
    "        merged_2['Headings'] = merged_2['Headings'].fillna('No')\n",
    "        merged_2['21_Conditions_list'] = merged_2['21_Conditions_list'].fillna('No')\n",
    "        merged_2['Gene_Score'] = merged_2['Gene_Score'].fillna('No')\n",
    "\n",
    "        print(\"Specific Genes Mapped\")\n",
    "        df_3 = pd.read_excel(r'new_final_output_concatenated.xlsx')\n",
    "\n",
    "        merged_2['POS'] = merged_2['POS'].astype('int64')\n",
    "        df_3['POS'] = df_3['POS'].astype('int64')\n",
    "\n",
    "        merged_2 = merged_2.rename({'Matched_Gene':'Gene'}, axis=1)\n",
    "\n",
    "        merged_3 = pd.merge(merged_2, df_3, on=['CHROM', 'POS', 'REF', 'ALT'], how='left', sort=False)\n",
    "        merged_3['Literature'] = merged_3['Literature'].fillna('No')\n",
    "\n",
    "        print(\"Lit Variants Mapped\")\n",
    "\n",
    "        merged_3 = merged_3[['Gene Name','Gene', 'Gene_Score', 'Condition', 'Headings', '21_Conditions_list', 'rsID', 'Literature', 'CHROM', 'POS', 'REF', 'ALT', 'Zygosity',\n",
    "            'Consequence', 'Consequence_score', 'IMPACT', 'IMPACT_score',\n",
    "            'ClinVar_CLNDN', 'CLIN_SIG', 'ClinVar_CLNREVSTAT', 'ClinVar', 'HGVSc',\n",
    "            'HGVSc (Transcript)', 'HGVSp', 'HGVSp (Transcript)', 'GT', 'GQ', 'SDP',\n",
    "            'DP', 'RD', 'AD', 'FREQ', 'PVAL', 'RDF', 'RDR', 'ADF', 'ADR', 'SIFT',\n",
    "            'PolyPhen', 'AF', 'AFR_AF', 'AMR_AF', 'EAS_AF', 'EUR_AF', 'SAS_AF',\n",
    "            'gnomADe_AF', 'gnomADe_AFR_AF', 'gnomADe_AMR_AF', 'gnomADe_ASJ_AF',\n",
    "            'gnomADe_EAS_AF', 'gnomADe_FIN_AF', 'gnomADe_NFE_AF', 'gnomADe_OTH_AF',\n",
    "            'gnomADe_SAS_AF', 'gnomADg_AF', 'gnomADg_AFR_AF', 'gnomADg_AMI_AF',\n",
    "            'gnomADg_AMR_AF', 'gnomADg_ASJ_AF', 'gnomADg_EAS_AF', 'gnomADg_FIN_AF',\n",
    "            'gnomADg_MID_AF', 'gnomADg_NFE_AF', 'gnomADg_OTH_AF', 'gnomADg_SAS_AF',\n",
    "            'MAX_AF', 'MAX_AF_POPS', 'BIOTYPE', 'EXON', 'INTRON',\n",
    "            'Protein Position and Amino Acid', 'Codons', 'STRAND', 'PUBMED']]\n",
    "\n",
    "        print(\"Exporting to excel\")\n",
    "        print('VCF processing Completed and Saved as Excel File')\n",
    "\n",
    "        print('Filter portion started')\n",
    "\n",
    "        condition_filter = merged_3[merged_3['21_Conditions_list'] != 'No']\n",
    "\n",
    "        consequence_filter = condition_filter[condition_filter['Consequence_score'].apply(lambda x: eval(x) >= 6/10)]\n",
    "\n",
    "        consequence_filter['DP'] = consequence_filter['DP'].replace('', '0').astype('int64')\n",
    "\n",
    "        dp_filter = consequence_filter[consequence_filter['DP'] >= 15]\n",
    "\n",
    "        dp_filter['gnomADe_AF'] = dp_filter['gnomADe_AF'].replace('', '0').astype(float)\n",
    "\n",
    "        gnomADe_AF_filter = dp_filter[dp_filter['gnomADe_AF'] <= 0.6]\n",
    "\n",
    "        gnomADe_AF_filter['gnomADe_SAS_AF'] = gnomADe_AF_filter['gnomADe_SAS_AF'].replace('', '0').astype(float)\n",
    "\n",
    "        gnomADe_SAS_AF_filter = gnomADe_AF_filter[gnomADe_AF_filter['gnomADe_SAS_AF'] <= 0.6]\n",
    "\n",
    "        df = gnomADe_SAS_AF_filter.copy()\n",
    "\n",
    "        print('Exon and Intron Filtering started')\n",
    "\n",
    "\n",
    "        # Replace empty strings with '0/0' and convert numerical parts to integers\n",
    "        df['EXON'] = df['EXON'].replace('', '0/0')\n",
    "        df['EXON_Numerator'] = df['EXON'].apply(lambda x: x.split('/')[0] if '/' in x else 0)\n",
    "        df['EXON_Denominator'] = df['EXON'].apply(lambda x: x.split('/')[1] if '/' in x else 0)\n",
    "\n",
    "        df['INTRON'] = df['INTRON'].replace('', '0/0')\n",
    "        df['INTRON_Numerator'] = df['INTRON'].apply(lambda x: x.split('/')[0] if '/' in x else 0)\n",
    "        df['INTRON_Denominator'] = df['INTRON'].apply(lambda x: x.split('/')[1] if '/' in x else 0)\n",
    "\n",
    "        # Convert the data types of numerator and denominator columns to integers\n",
    "        df['EXON_Numerator'] = df['EXON_Numerator'].astype(int)\n",
    "        df['EXON_Denominator'] = df['EXON_Denominator'].astype(int)\n",
    "\n",
    "        df['INTRON_Numerator'] = df['INTRON_Numerator'].astype(int)\n",
    "        df['INTRON_Denominator'] = df['INTRON_Denominator'].astype(int)\n",
    "\n",
    "        # Initialize an empty DataFrame to store the final result\n",
    "        result_df = pd.DataFrame()\n",
    "\n",
    "        # Iterate over unique combinations of CHROM, POS, rsID, REF, ALT\n",
    "        for _, group_df in df.groupby(['CHROM', 'POS', 'REF', 'ALT']):\n",
    "            # Check if EXON column has values\n",
    "            if not group_df['EXON_Numerator'].eq(0).all():\n",
    "                # Prioritize rows with values in EXON column\n",
    "                result_df = pd.concat([result_df, group_df.sort_values(by=['EXON_Numerator'], ascending=False).head(1)])\n",
    "            else:\n",
    "                # If EXON is empty, prioritize rows with values in INTRON column\n",
    "                if not group_df['INTRON_Numerator'].eq(0).all():\n",
    "                    result_df = pd.concat([result_df, group_df.sort_values(by=['INTRON_Numerator'], ascending=False).head(1)])\n",
    "                else:\n",
    "                    # If both EXON and INTRON are empty, just concatenate the first row\n",
    "                    result_df = pd.concat([result_df, group_df.head(1)])\n",
    "\n",
    "        # Now result_df contains the rows you're looking for based on the specified logic\n",
    "        result_df = result_df.drop(['EXON_Numerator', 'EXON_Denominator', 'INTRON_Numerator', 'INTRON_Denominator'], axis=1)\n",
    "\n",
    "        result_df['EXON'] = result_df['EXON'].astype('str')\n",
    "        result_df['INTRON'] = result_df['INTRON'].astype('str')\n",
    "                                        \n",
    "        ########################################################################################################################\n",
    "\n",
    "        drop_duplicates_filter = result_df.copy()\n",
    "\n",
    "        drop_duplicates_filter = drop_duplicates_filter.set_index(['Gene Name', 'Gene', 'Gene_Score', 'rsID', 'Literature', 'CHROM', 'POS', 'REF',\n",
    "            'ALT', 'Zygosity', 'Consequence', 'Consequence_score', 'IMPACT',\n",
    "            'IMPACT_score', 'ClinVar_CLNDN', 'CLIN_SIG', 'ClinVar_CLNREVSTAT',\n",
    "            'ClinVar', 'HGVSc', 'HGVSc (Transcript)', 'HGVSp', 'HGVSp (Transcript)',\n",
    "            'GT', 'GQ', 'SDP', 'DP', 'RD', 'AD', 'FREQ', 'PVAL', 'RDF', 'RDR',\n",
    "            'ADF', 'ADR', 'SIFT', 'PolyPhen', 'AF', 'AFR_AF', 'AMR_AF', 'EAS_AF',\n",
    "            'EUR_AF', 'SAS_AF', 'gnomADe_AF', 'gnomADe_AFR_AF', 'gnomADe_AMR_AF',\n",
    "            'gnomADe_ASJ_AF', 'gnomADe_EAS_AF', 'gnomADe_FIN_AF', 'gnomADe_NFE_AF',\n",
    "            'gnomADe_OTH_AF', 'gnomADe_SAS_AF', 'gnomADg_AF', 'gnomADg_AFR_AF',\n",
    "            'gnomADg_AMI_AF', 'gnomADg_AMR_AF', 'gnomADg_ASJ_AF', 'gnomADg_EAS_AF',\n",
    "            'gnomADg_FIN_AF', 'gnomADg_MID_AF', 'gnomADg_NFE_AF', 'gnomADg_OTH_AF',\n",
    "            'gnomADg_SAS_AF', 'MAX_AF', 'MAX_AF_POPS', 'BIOTYPE', 'EXON', 'INTRON',\n",
    "            'Protein Position and Amino Acid', 'Codons', 'STRAND', 'PUBMED']).apply(lambda x: x.str.split('; ').explode()).reset_index()\n",
    "        # Specify the column names to move to the beginning\n",
    "        columns_to_move = ['Condition', 'Headings', '21_Conditions_list']\n",
    "\n",
    "        # Reorder the DataFrame to move specified columns to the beginning\n",
    "        drop_duplicates_filter = drop_duplicates_filter[columns_to_move + [col for col in drop_duplicates_filter.columns if col not in columns_to_move]]\n",
    "        search = re.search(r'(\\d+)', filename) # looks for numerical digits in filename, in our case that is the patient id\n",
    "        patientId = search.group(1) \n",
    "        drop_duplicates_filter.insert(loc=0, column='patient_id', value=patientId)\n",
    "        print('filtering completed')\n",
    "\n",
    "        #######################################################################################################################################\n",
    "\n",
    "        drop_duplicates_filter.to_csv(f\"{output_dir}/{filename}.csv\",index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d17eac60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading the Data completed and Depth columns splitted\n",
      "Zygosity extraction completed\n",
      "Gene extraction completed\n",
      "CSQ splitting completed\n",
      "Required columns extraction completed\n",
      "Protein_HGVSc_HGVSp_extraction completed\n",
      "\"not_specified\", \"not_provided\" completed\n",
      "Scores added\n",
      "Specific Genes Mapped\n",
      "Lit Variants Mapped\n",
      "Exporting to excel\n",
      "VCF processing Completed and Saved as Excel File\n",
      "Filter portion started\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: 'nan'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mprocess\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../vcfFiles/IIIT_HYD_Anonymous39.vcf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m../processed\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[4], line 237\u001b[0m, in \u001b[0;36mprocess\u001b[0;34m(filepath, output_dir)\u001b[0m\n\u001b[1;32m    233\u001b[0m condition_filter \u001b[38;5;241m=\u001b[39m merged_3[merged_3[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m21_Conditions_list\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNo\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m    235\u001b[0m consequence_filter \u001b[38;5;241m=\u001b[39m condition_filter[condition_filter[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mConsequence_score\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mapply(\u001b[38;5;28;01mlambda\u001b[39;00m x: \u001b[38;5;28meval\u001b[39m(x) \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m6\u001b[39m\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m10\u001b[39m)]\n\u001b[0;32m--> 237\u001b[0m consequence_filter[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDP\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43mconsequence_filter\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mDP\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreplace\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m0\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mastype\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mint64\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    239\u001b[0m dp_filter \u001b[38;5;241m=\u001b[39m consequence_filter[consequence_filter[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDP\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m15\u001b[39m]\n\u001b[1;32m    241\u001b[0m dp_filter[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgnomADe_AF\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m dp_filter[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgnomADe_AF\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m0\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mfloat\u001b[39m)\n",
      "File \u001b[0;32m~/genepower/.venv/lib/python3.12/site-packages/pandas/core/generic.py:6637\u001b[0m, in \u001b[0;36mNDFrame.astype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m   6631\u001b[0m     results \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m   6632\u001b[0m         ser\u001b[38;5;241m.\u001b[39mastype(dtype, copy\u001b[38;5;241m=\u001b[39mcopy, errors\u001b[38;5;241m=\u001b[39merrors) \u001b[38;5;28;01mfor\u001b[39;00m _, ser \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mitems()\n\u001b[1;32m   6633\u001b[0m     ]\n\u001b[1;32m   6635\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   6636\u001b[0m     \u001b[38;5;66;03m# else, only a single dtype is given\u001b[39;00m\n\u001b[0;32m-> 6637\u001b[0m     new_data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_mgr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mastype\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6638\u001b[0m     res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_constructor_from_mgr(new_data, axes\u001b[38;5;241m=\u001b[39mnew_data\u001b[38;5;241m.\u001b[39maxes)\n\u001b[1;32m   6639\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m res\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mastype\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/genepower/.venv/lib/python3.12/site-packages/pandas/core/internals/managers.py:431\u001b[0m, in \u001b[0;36mBaseBlockManager.astype\u001b[0;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[1;32m    428\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m using_copy_on_write():\n\u001b[1;32m    429\u001b[0m     copy \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[0;32m--> 431\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    432\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mastype\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    433\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    434\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    435\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    436\u001b[0m \u001b[43m    \u001b[49m\u001b[43musing_cow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43musing_copy_on_write\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    437\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/genepower/.venv/lib/python3.12/site-packages/pandas/core/internals/managers.py:364\u001b[0m, in \u001b[0;36mBaseBlockManager.apply\u001b[0;34m(self, f, align_keys, **kwargs)\u001b[0m\n\u001b[1;32m    362\u001b[0m         applied \u001b[38;5;241m=\u001b[39m b\u001b[38;5;241m.\u001b[39mapply(f, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    363\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 364\u001b[0m         applied \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mf\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    365\u001b[0m     result_blocks \u001b[38;5;241m=\u001b[39m extend_blocks(applied, result_blocks)\n\u001b[1;32m    367\u001b[0m out \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m)\u001b[38;5;241m.\u001b[39mfrom_blocks(result_blocks, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxes)\n",
      "File \u001b[0;32m~/genepower/.venv/lib/python3.12/site-packages/pandas/core/internals/blocks.py:758\u001b[0m, in \u001b[0;36mBlock.astype\u001b[0;34m(self, dtype, copy, errors, using_cow, squeeze)\u001b[0m\n\u001b[1;32m    755\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCan not squeeze with more than one column.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    756\u001b[0m     values \u001b[38;5;241m=\u001b[39m values[\u001b[38;5;241m0\u001b[39m, :]  \u001b[38;5;66;03m# type: ignore[call-overload]\u001b[39;00m\n\u001b[0;32m--> 758\u001b[0m new_values \u001b[38;5;241m=\u001b[39m \u001b[43mastype_array_safe\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    760\u001b[0m new_values \u001b[38;5;241m=\u001b[39m maybe_coerce_values(new_values)\n\u001b[1;32m    762\u001b[0m refs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/genepower/.venv/lib/python3.12/site-packages/pandas/core/dtypes/astype.py:237\u001b[0m, in \u001b[0;36mastype_array_safe\u001b[0;34m(values, dtype, copy, errors)\u001b[0m\n\u001b[1;32m    234\u001b[0m     dtype \u001b[38;5;241m=\u001b[39m dtype\u001b[38;5;241m.\u001b[39mnumpy_dtype\n\u001b[1;32m    236\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 237\u001b[0m     new_values \u001b[38;5;241m=\u001b[39m \u001b[43mastype_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    238\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mValueError\u001b[39;00m, \u001b[38;5;167;01mTypeError\u001b[39;00m):\n\u001b[1;32m    239\u001b[0m     \u001b[38;5;66;03m# e.g. _astype_nansafe can fail on object-dtype of strings\u001b[39;00m\n\u001b[1;32m    240\u001b[0m     \u001b[38;5;66;03m#  trying to convert to float\u001b[39;00m\n\u001b[1;32m    241\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m errors \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m~/genepower/.venv/lib/python3.12/site-packages/pandas/core/dtypes/astype.py:182\u001b[0m, in \u001b[0;36mastype_array\u001b[0;34m(values, dtype, copy)\u001b[0m\n\u001b[1;32m    179\u001b[0m     values \u001b[38;5;241m=\u001b[39m values\u001b[38;5;241m.\u001b[39mastype(dtype, copy\u001b[38;5;241m=\u001b[39mcopy)\n\u001b[1;32m    181\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 182\u001b[0m     values \u001b[38;5;241m=\u001b[39m \u001b[43m_astype_nansafe\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    184\u001b[0m \u001b[38;5;66;03m# in pandas we don't store numpy str dtypes, so convert to object\u001b[39;00m\n\u001b[1;32m    185\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(dtype, np\u001b[38;5;241m.\u001b[39mdtype) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28missubclass\u001b[39m(values\u001b[38;5;241m.\u001b[39mdtype\u001b[38;5;241m.\u001b[39mtype, \u001b[38;5;28mstr\u001b[39m):\n",
      "File \u001b[0;32m~/genepower/.venv/lib/python3.12/site-packages/pandas/core/dtypes/astype.py:133\u001b[0m, in \u001b[0;36m_astype_nansafe\u001b[0;34m(arr, dtype, copy, skipna)\u001b[0m\n\u001b[1;32m    129\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[1;32m    131\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m copy \u001b[38;5;129;01mor\u001b[39;00m arr\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mobject\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m dtype \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mobject\u001b[39m:\n\u001b[1;32m    132\u001b[0m     \u001b[38;5;66;03m# Explicit copy, or required since NumPy can't view from / to object.\u001b[39;00m\n\u001b[0;32m--> 133\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43marr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mastype\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m    135\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m arr\u001b[38;5;241m.\u001b[39mastype(dtype, copy\u001b[38;5;241m=\u001b[39mcopy)\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: 'nan'"
     ]
    }
   ],
   "source": [
    "process(\"../vcfFiles/IIIT_HYD_Anonymous39.vcf\",\"../processed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
